tensorflow/core/framework/op_kernel.cc

```diff
@@ -1276,6 +1279,24 @@ void OpKernelRegistrar::InitInternal(const KernelDef* kernel_def,
   delete kernel_def;
 }
 
+void OpKernelRegistrar::Register(const KernelDef* kernel_def,
```

Comment:
I don't think we should duplicate the logic from `OpKernelRegistrar::InitInternal`.  Can we instead use `kernel_def->subdevice_type()` in the key in both places?  Also see other comment about subdevice type for internal kernels.

Reply:
@sanjoy, We thought that OpKernelRegistrar::InitInternal is for those static registered kernels in the proper, and Register is for those dynamic registered kernels from the plugins, and for those static registered kernels(existing kernels in the proper), we can set treat the device_type as the subdevice_type for the kernel registration.
Currently, existing kernels in the tensorflow proper are registered through REGISTER_KERNEL_BUILDER(), and it didn't set the subdevice attribute in the kernel_def. if we want to use the subdevice type as the key of registeration, we need to change all REGISTER_KERNEL_BUILDER(), I think that change is too huge.
```
 REGISTER_KERNEL_BUILDER(Name("Save").Device(DEVICE_CPU), SaveOp);
->
 REGISTER_KERNEL_BUILDER(Name("Save").Device(DEVICE_CPU).SubDevice(DEVICE_CPU), SaveOp);
REGISTER_KERNEL_BUILDER(Name("Reshape").Device(DEVICE_GPU).HostMemory("shape"), ReshapeOp);
->
REGISTER_KERNEL_BUILDER(Name("Reshape").Device(DEVICE_GPU).SubDevice(DEVICE_GPU).HostMemory("shape"), ReshapeOp);
```

Reply:
done, I change the subdevice type as empty string for those first party devices and as the kernel_def.proto will make the default string as empty, so even we don't explicit specify the subdevice string for the first party devices, it will be ok.
Thanks.